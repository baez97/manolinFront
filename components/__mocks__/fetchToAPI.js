const fetchToAPI = jest.fn(() => new Promise((res, rej) => 
    res({
        json: () => ({mocked: "mocked"}) 
    })
));
export default fetchToAPI;