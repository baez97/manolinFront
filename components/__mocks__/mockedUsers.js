const documents = [
    {
        "_id": {
            "$oid": "5c7ab0c331f6240fbc248f6e"
        },
        "name": "Ana",
        "password": "$2a$08$Er1/dLfJSdvJojynt6Ykj.UvK..Hd.9P1Cs7Akkyb0QdX92pEXePm",
        "email": null,
        "turnWithDates": [
            {
                "day": 1,
                "month": 1,
                "year": 2019,
                "weekday": 2,
                "turn": "M"
            },
            {
                "day": 2,
                "month": 1,
                "year": 2019,
                "weekday": 3,
                "turn": "M"
            },
            {
                "day": 3,
                "month": 1,
                "year": 2019,
                "weekday": 4,
                "turn": "T"
            },
            {
                "day": 4,
                "month": 1,
                "year": 2019,
                "weekday": 5,
                "turn": "T"
            }
        ]
    },
    {
        "_id": {
            "$oid": "5c7c0e8a5434ee0287d3e6b6"
        },
        "name": "Mar√≠a",
        "password": "$2a$08$W2mbg6LEEIxx81h9p9VFj.RF6UASLiou/bLQul7y3ITgFhwbdNmWC",
        "email": null,
        "turnWithDates": [
            {
                "day": 1,
                "month": 1,
                "year": 2019,
                "weekday": 2,
                "turn": "M"
            },
            {
                "day": 2,
                "month": 1,
                "year": 2019,
                "weekday": 3,
                "turn": "M"
            },
            {
                "day": 3,
                "month": 1,
                "year": 2019,
                "weekday": 4,
                "turn": "T"
            }
        ]
    }
];

module.exports = documents;