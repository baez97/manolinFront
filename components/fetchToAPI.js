import { BACKEND_IP } from '../config';

function fetchToAPI(urlString, options) {
    return fetch( BACKEND_IP + urlString, options);
}

export default fetchToAPI;