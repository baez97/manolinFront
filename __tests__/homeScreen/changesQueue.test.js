import React from 'react';
import renderer from 'react-test-renderer';
import ChangesQueue from '../../components/homeComponents/changes/changesQueue';
import fetchToAPI from '../../components/fetchToAPI';
import mockedUsers from '../../components/__mocks__/mockedUsers';

jest.unmock("react-native");
jest.unmock("react");
jest.unmock("react-test-renderer");
jest.unmock("expo");
jest.unmock("../../components/homeComponents/changes/changeView")
jest.mock("socket.io-client");
jest.mock("../../components/fetchToAPI");


const simulateRenderItem = function(listContainer, item) {
    listContainer
        .props
        .children
        .props
        .renderItem({item});
}

const simulateSetState = function(reactObject, stateObj) {
    reactObject
        .state = stateObj;
}

describe("ChangesQueue", () => {
    
    const mockedToken = "TOKEN";
    const props = {
        token: mockedToken,
        socket: {
            on: jest.fn((chain, callback) => {
                callback()
            }),
            emit: jest.fn()
        },
        onLoaded: jest.fn()
    };
    
    fetchToAPI.mockReturnValueOnce(
        new Promise((res, rej) => {
            res(mockedUsers[0])
        })
    );

    const cQ = new ChangesQueue(props);
    cQ.setState   = jest.fn();

    describe("ComponentDidMount", () => {

        it("Calls to fetchToAPI", async() => {
            fetchToAPI.mockReturnValueOnce(
                new Promise((res, rej) => {
                    res({json: () => ({})})
                }));
            await cQ.componentDidMount();
            expect( fetchToAPI ).toBeCalled();
        });

        it("Sets the state correctly when there is not an error", async () => {
            const mockedResponse = {err: false, result: "MOCKEDRESULT"};
            const expectedArgs = {
                changesLoaded: true,
                changes: mockedResponse.result
            }
            fetchToAPI.mockReturnValueOnce(
                new Promise((res, rej) => {
                    res({json: () => (mockedResponse)})
                }));
            await cQ.componentDidMount();
            expect( cQ.setState ).toBeCalled();
            expect( cQ.setState ).toBeCalledWith(expectedArgs)
        });

        it("Sets the state correctly when there is an error", async () => {
            const mockedResponse = {err: true, result: []};
            const expectedArgs = {
                changesLoaded: true,
                changes: mockedResponse.result
            }
            fetchToAPI.mockReturnValueOnce(
                new Promise((res, rej) => {
                    res({json: () => (mockedResponse)})
                }));
            await cQ.componentDidMount();
            expect( cQ.setState ).toBeCalled();
            expect( cQ.setState ).toBeCalledWith(expectedArgs)
        });

        it("Sets the state correctly when an error is thrown", async () => {
            const expectedArgs = {
                changesLoaded: true,
                changes: []
            }
            fetchToAPI.mockReturnValueOnce(
                new Promise((res, rej) => {
                    throw "ERR";
                }));
            await cQ.componentDidMount();
            expect( cQ.setState ).toBeCalled();
            expect( cQ.setState ).toBeCalledWith(expectedArgs)
        });
    });

    describe("View", () => {
        it("Matches the Snapshot when changes are not loaded yet", () => {
            const view = <ChangesQueue token={mockedToken} socket={{on: jest.fn()}}/>
            const rendered = renderer.create(view);
            const renderedJson = rendered.toJSON();

            expect( renderedJson ).toMatchSnapshot();
        });

        it("Matches the Snapshot when there are no changes", async () => {
            const view     = <ChangesQueue token="TOKEN" socket={{on: jest.fn()}}/>
            const rendered = await renderer.create(view);
            const renderedObj = rendered.root._fiber.stateNode;
            renderedObj.setState({
                changesLoaded: true,
                changes: []
            });

            const renderedJson = rendered.toJSON();
            
            expect ( renderedJson ).toMatchSnapshot();
        });

        it("RenderItem works correctly", async () => {
            const mockedChange = {
                _id: "id_1",
                day: 1,
                month: 1,
                year: 2019,
                turn: 'M',
                weekday: 2
            }

            simulateSetState( cQ, {
                changesLoaded:true,
                changes: [Â mockedChange ]
            });

            const rendered = cQ.render();
            React.createElement = jest.fn();
            simulateRenderItem( rendered, mockedChange );
            expect( React.createElement ).toBeCalled();
        });

        it("KeyExtractor works correctly", () => {
            const expectedOutput = "id_1";
            const mockedChange = {
                _id: "id_1",
                day: 1,
                month: 1,
                year: 2019,
                turn: 'M',
                weekday: 2
            }

            const actualOutput = cQ.keyExtractor(mockedChange);
            expect( actualOutput ).toBe(expectedOutput);
        })


    });
});