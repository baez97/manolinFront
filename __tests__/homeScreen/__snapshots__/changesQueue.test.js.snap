// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChangesQueue View Matches the Snapshot when changes are not loaded yet 1`] = `
<Text>
  Cargando cambios...
</Text>
`;

exports[`ChangesQueue View Matches the Snapshot when there are no changes 1`] = `<View />`;
